be/src/olap/rowset/segment_v2/binary_dict_page.cpp

```diff
@@ -0,0 +1,223 @@
+                if (item_mem == nullptr) {
+                    return Status::Corruption(Substitute("memory allocate failed, size:$0", src->size));
+                }
+                Slice dict_item(src->data, src->size);
+                dict_item.relocate(item_mem);
+                value_code = _dictionary.size();
+                _dictionary.insert({dict_item, value_code});
+                _dict_items.push_back(dict_item);
+                _dict_builder->update_prepared_size(dict_item.size);
+            }
+            RETURN_IF_ERROR(_data_page_builder->add(reinterpret_cast<const uint8_t*>(&value_code), &add_count));
+            if (add_count == 0) {
+                // current data page is full, stop processing remaining inputs
+                break;
+            }
+            num_added += 1;
+        }
+        *count = num_added;
+        return Status::OK();
+    } else {
+        DCHECK_EQ(_encoding_type, PLAIN_ENCODING);
+        return _data_page_builder->add(vals, count);
+    }
+}
+
+Slice BinaryDictPageBuilder::finish() {
+    _finished = true;
+
+    Slice data_slice = _data_page_builder->finish();
+    _buffer.append(data_slice.data, data_slice.size);
+    encode_fixed32_le(&_buffer[0], _encoding_type);
+    return Slice(_buffer.data(), _buffer.size());
+}
+
+void BinaryDictPageBuilder::reset() {
+    _finished = false;
+    _buffer.reserve(_options.data_page_size + BINARY_DICT_PAGE_HEADER_SIZE);
+    _buffer.resize(BINARY_DICT_PAGE_HEADER_SIZE);
+
+    if (_encoding_type == DICT_ENCODING
+            && _dict_builder->is_page_full()) {
+        _data_page_builder.reset(new BinaryPlainPageBuilder(_options));
+        _encoding_type = PLAIN_ENCODING;
+    } else {
+        _data_page_builder->reset();
+    }
+    _finished = false;
+}
+
+size_t BinaryDictPageBuilder::count() const {
+    return _data_page_builder->count();
+}
+
+Status BinaryDictPageBuilder::get_dictionary_page(Slice* dictionary_page) {
+    DCHECK(_finished) << "get dictionary page when the builder is not finished";
+    _dictionary.clear();
+    _dict_builder->reset();
+    size_t add_count = 1;
+    // here do not check is_page_full of dict_builder
+    // because it is checked in add
+    for (auto& dict_item: _dict_items) {
+        RETURN_IF_ERROR(_dict_builder->add(reinterpret_cast<const uint8_t*>(&dict_item), &add_count));
+    }
+    *dictionary_page = _dict_builder->finish();
+    _dict_builder->release();
+    _dict_items.clear();
+    return Status::OK();
+}
+
+BinaryDictPageDecoder::BinaryDictPageDecoder(Slice data, const PageDecoderOptions& options) :
+    _data(data),
+    _options(options),
+    _data_page_decoder(nullptr),
+    _dict_decoder(options.dict_decoder),
+    _parsed(false),
+    _encoding_type(UNKNOWN_ENCODING) { }
+
+Status BinaryDictPageDecoder::init() {
+    CHECK(!_parsed);
+    if (_data.size < BINARY_DICT_PAGE_HEADER_SIZE) {
+        LOG(WARNING) << "corrupted data, data size:" << _data.size;
```

Comment:
I think it's better not to print log here, let client print log

Reply:
ok